plugins/out_s3/s3.c

```diff
@@ -1066,10 +1107,18 @@ static int upload_data(struct flb_s3 *ctx, struct s3_file *chunk,
     if (ret < 0) {
         /* re-add chunk to list */
         if (chunk) {
-            s3_store_file_unlock(chunk);
             chunk->failures += 1;
+            if (chunk->failures > ctx->ins->retry_limit){
+                s3_retry_warn(ctx, tag, chunk->input_name, create_time, FLB_FALSE);
+                s3_store_file_delete(ctx, chunk);
+                return FLB_ERROR;
+            }
```

Comment:
I'm pretty sure the code will segfault if preserve_data_ordering is turned on, because we are not removing the s3_file from the upload queue.

Have you tested with preserve_data_ordering option?

Reply:
S3 upload queue calls out to upload_data in order to do the upload... so the s3 upload queue should get the error return here and remove from the queue. That's my memory of how it works. But year Clay you should read it and understand for yourself and answer matt here.